index.rst

```diff
@@ -48,16 +48,301 @@
+  var db = 'squash-prod'
+  var rp = 'autogen'
+  var measurement = 'ap_association'
+  var groupBy = ['visit', 'ccdnum', 'ci_dataset']
+  var whereFilter = lambda: TRUE
+
+  var message = '{{.Name}} is {{.Level}} on build #{{ index .Tags "ci_id" }}: AssociationTime = {{ index .Fields "value" | printf "%0.2f s" }} for {{.Group}}'
+
+  var triggerType = 'threshold'
+  var crit = 5
+
+  var data = stream
+      |from()
+          .database(db)
+          .retentionPolicy(rp)
+          .measurement(measurement)
+          .groupBy(groupBy)
+          .where(whereFilter)
+      |eval(lambda: "ap_association.AssociationTime")
+          .as('value')
+  var trigger = data
+      |alert()
+          .crit(lambda: "value" > crit)
+          .message(message)
+          .stateChangesOnly()
+          .slack()
+          .channel('#dm-squash-alerts')
+
+Example of a notification message produced by this task:
+
+    *ap_association is CRITICAL on build #279:
+    AssociationTime = 5.42s for ccdnum=56, ci_dataset=CI-HiTS2015, visit=411371*
+
+
+|34| suggests a “subscription list” for each metric to be defined, and the key stakeholders automatically be added to it for all metrics deriving directly from high-level requirements. This could be achieved by sending notifications to specific slack channels for example, notification about regression on AP metrics are sent to ``#dm-alert-prod``, notifications about regression on DRP metrics to ``#dm-drp``, etc.
+
+
+Supporting multiple execution environments
+==========================================
+
+To be generally useful for the verification activities, SQuaSH must support multiple execution environments.
+
+The following project environments are currently supported:
+
+* Jenkins CI
+* LDF
+
+SQuaSH captures environment variables from these environments and use them as metadata associated with the metric values.
+
+.. todo:: Document the required environment variables in each situation and the corresponding metadata tags used by SQuaSH.
+
+SQuaSH has the concept of runs. A run may contain results from several verification jobs executed on a given environment. For example a ``GET`` request to ``/jenkins/<run_id>`` or to ``/lfd/<run_id>`` will retrieve all the verification jobs in that run.
+
+In DM-18505_, we add support for a local execution environment.  Adding support to a local execution environment, allows DM developers to run verification jobs in the notebook aspect of the LSP or from their laptop and dispatch results to SQuaSH. This implementation fulfills |37|.
+
+.. note::
+
+  Dispatching results to SQuaSH requires auth access to the SQuaSH API. Currently, the only mechanism to register new users is interacting to the SQuaSH API. That can be simplified using the GitHub OAuth2 provider in SQuaSH .
+
+.. note::
+
+  Because the local execution environment is not a controlled environment like the Jenkins CI or the LDF, we can not capture information such as code version or configuration.
```

Comment:
How will we support multiple users?  Will the local execution environment forward the `$USER` env var, or will we have multiple execution environments, one per user?

Reply:
My plan is to implement one execution environment per user, with that we can have in SQuaSH incremental run IDs per user which is good. ``env_name`` is going to be the username registered in SQuaSH.